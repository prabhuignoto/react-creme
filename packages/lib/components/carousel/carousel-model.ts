import React from 'react';

export type CarouselProps = {
  autoPlay?: number;
  border?: boolean;
  children: React.ReactNode | React.ReactNode[];
  direction?: 'horizontal' | 'vertical';
  enableSwipe?: boolean;
  focusable?: boolean;
  height?: number;
  size?: 'sm' | 'md' | 'lg';
  transition?: string;
};

export type CarouselItemProps = {
  height: number;
  id: string;
  left?: string;
  top?: string;
  visible?: boolean;
  width: number;
};

export type CarouselItemsProps = Pick<
  CarouselProps,
  'children' | 'direction' | 'height' | 'size'
> & {
  activePage: number;
  carouselItems: CarouselItemProps[];
  totalItems: number;
  width: number;
};

export type CarouselTrackProps = Pick<
  CarouselProps,
  'direction' | 'focusable' | 'size'
> & {
  activeIndex: number;
  focusable?: boolean;
  handleSelection: (index: number) => void;
  hideNext?: boolean;
  hidePrevious?: boolean;
  length: number;
  onNext: () => void;
  onPrevious: () => void;
};

export type CarouselButtonProps = Pick<
  CarouselProps,
  'direction' | 'focusable' | 'size'
> & {
  hide?: boolean;
  label: string;
  onClick: () => void;
  position: 'left' | 'right';
};
