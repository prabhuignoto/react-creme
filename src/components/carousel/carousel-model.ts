import React from "react";

export interface CarouselModel {
  children: React.ReactNode | React.ReactNode;
  direction?: "horizontal" | "vertical";
  height?: number;
}

export interface CarouselItemModel {
  height?: number;
  id: string;
  left?: string;
  top?: string;
  visible?: boolean;
  width?: number;
}

export interface CarouselItemsModel {
  activePage: number;
  carouselItems: CarouselItemModel[];
  children: React.ReactNode[] | React.ReactNode;
  direction: "horizontal" | "vertical";
  height: number;
  totalItems: number;
  width: number;
}
