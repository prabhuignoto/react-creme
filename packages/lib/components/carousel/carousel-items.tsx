import React, { CSSProperties, ReactNode, useMemo } from "react";
import { CarouselItemsModel } from "./carousel-model";
import "./carousel.scss";

const CarouselItems: React.FunctionComponent<CarouselItemsModel> = ({
  activePage = 0,
  carouselItems,
  children,
  direction,
  height,
  totalItems,
  width,
}: CarouselItemsModel) => {
  const carouselStyle = useMemo(() => {
    return {
      width: `${width * totalItems}px`,
      transform:
        direction === "horizontal"
          ? `translateX(-${width * activePage}px)`
          : `translateY(-${height * activePage}px)`,
    } as CSSProperties;
  }, [width, height, activePage]);

  return (
    <ul className={"rc-carousel"} style={carouselStyle} role="list">
      {carouselItems.map((item, index) => (
        <li
          key={item.id}
          className={"rc-carousel-item"}
          role="listitem"
          data-visible={activePage === index}
          aria-hidden={activePage !== index}
          style={{
            width: item.width,
            visibility: item.width === 0 ? "hidden" : "visible",
            top: direction === "horizontal" ? 0 : item.top,
            left: direction === "horizontal" ? item.left : 0,
          }}
        >
          <div className="rc-carousel-item-container">
            {(children as ReactNode[])[index]}
          </div>
        </li>
      ))}
    </ul>
  );
};

CarouselItems.displayName = "CarouselItems";

export { CarouselItems };
