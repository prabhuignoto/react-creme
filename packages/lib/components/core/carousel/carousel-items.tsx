import React, { CSSProperties, ReactNode, useMemo } from 'react';
import { CarouselItemsProps } from './carousel-model';
import styles from './carousel.module.scss';

const CarouselItems: React.FunctionComponent<CarouselItemsProps> = ({
  activePage = 0,
  carouselItems,
  children,
  direction,
  height,
  totalItems,
  width,
}: CarouselItemsProps) => {
  const carouselStyle = useMemo(() => {
    return {
      transform:
        direction === 'horizontal'
          ? `translateX(-${width * activePage}px)`
          : `translateY(-${height * activePage}px)`,
      width: `${width * totalItems}px`,
    } as CSSProperties;
  }, [width, height, activePage]);

  return (
    <ul className={styles.carousel} style={carouselStyle} role="list">
      {carouselItems.map((item, index) => (
        <li
          key={item.id}
          className={styles.item}
          role="listitem"
          data-visible={activePage === index}
          aria-hidden={activePage !== index}
          style={{
            left: direction === 'horizontal' ? item.left : 0,
            top: direction === 'horizontal' ? 0 : item.top,
            visibility: item.width === 0 ? 'hidden' : 'visible',
            width: item.width,
          }}
        >
          <div className={styles.item_container}>
            {(children as ReactNode[])[index]}
          </div>
        </li>
      ))}
    </ul>
  );
};

CarouselItems.displayName = 'CarouselItems';

export { CarouselItems };
