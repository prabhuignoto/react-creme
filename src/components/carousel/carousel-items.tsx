import React, { CSSProperties, ReactNode, useMemo } from "react";
import { CarouselItemsModel } from "./carousel-model";
import "./carousel.scss";

const CarouselItems: React.FunctionComponent<CarouselItemsModel> = ({
  activePage = 0,
  carouselItems,
  children,
  direction,
  height,
  totalItems,
  width,
}: CarouselItemsModel) => {
  const carouselStyle = useMemo(() => {
    return {
      width: `${width * totalItems}px`,
      transform:
        direction === "horizontal"
          ? `translateX(-${width * activePage}px)`
          : `translateY(-${height * activePage}px)`,
    } as CSSProperties;
  }, [width, height, activePage]);

  return (
    <ul className={"rc-carousel"} style={carouselStyle} role="list">
      {carouselItems.map((item, index) => (
        <li
          key={item.id}
          className={"rc-carousel-item"}
          role="listitem"
          style={{
            width: item.width,
            height: direction === "horizontal" ? "100%" : item.height - 32,
            top: direction === "horizontal" ? 0 : item.top,
            left: direction === "horizontal" ? item.left : 0,
          }}
        >
          <div className="rc-carousel-item-container">
            {(children as ReactNode[])[index]}
          </div>
        </li>
      ))}
    </ul>
  );
};

CarouselItems.displayName = "CarouselItems";

export { CarouselItems };
