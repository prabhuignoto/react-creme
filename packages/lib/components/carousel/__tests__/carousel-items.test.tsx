import { render } from '@testing-library/react';
import { CarouselItems } from '../carousel-items';
import { CarouselItemProps } from './../carousel-model';

const items: CarouselItemProps[] = [
  {
    height: 300,
    id: '234',
    left: '0px',
    top: '0px',
    visible: true,
    width: 300,
  },
  {
    height: 300,
    id: '235',
    left: '300px',
    top: '0px',
    visible: true,
    width: 300,
  },
];

describe('Carousel items', () => {
  it('Should render carousel items', async () => {
    const { getByRole, getAllByRole } = render(
      <CarouselItems
        carouselItems={items}
        height={300}
        width={300}
        direction="horizontal"
        totalItems={2}
        activePage={1}
      >
        <span>one</span>
        <span>two</span>
      </CarouselItems>,
      {
        container: document.body,
      }
    );
    expect(getByRole('list')).toBeInTheDocument();
    expect(getAllByRole('listitem')).toHaveLength(1);
  });

  it('Should render carousel items snapshot', async () => {
    const { getByRole } = render(
      <CarouselItems
        carouselItems={items}
        height={300}
        width={300}
        direction="horizontal"
        totalItems={2}
        activePage={1}
      >
        <span>one</span>
        <span>two</span>
      </CarouselItems>,
      {
        container: document.body,
      }
    );
    expect(getByRole('list')).toMatchSnapshot();
  });
});
