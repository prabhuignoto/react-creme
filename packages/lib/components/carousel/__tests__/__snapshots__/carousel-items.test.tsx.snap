// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel items Should render carousel items snapshot 1`] = `
<ul
  class="rc-carousel"
  role="list"
  style="width: 600px; transform: translateX(-300px);"
>
  <li
    class="rc-carousel-item"
    data-visible="false"
    role="listitem"
    style="width: 300px; visibility: visible; top: 0px; left: 0px;"
  >
    <div
      class="rc-carousel-item-container"
    >
      <span>
        one
      </span>
    </div>
  </li>
  <li
    class="rc-carousel-item"
    data-visible="true"
    role="listitem"
    style="width: 300px; visibility: visible; top: 0px; left: 300px;"
  >
    <div
      class="rc-carousel-item-container"
    >
      <span>
        two
      </span>
    </div>
  </li>
</ul>
`;
